packages/react-reconciler/src/ReactFiberHydrationContext.js

```diff
@@ -99,21 +99,60 @@ function deleteHydratableInstance(
   }
 }
 
-function insertNonHydratedInstance(returnFiber: Fiber, fiber: Fiber) {
+function insertNonHydratedInstance(
+  returnFiber: Fiber,
+  fiber: Fiber,
+  hydrationWarningHostInstanceIsReplaced: boolean,
+) {
   fiber.effectTag |= Placement;
   if (__DEV__) {
+    let hydrationWarningHostInstanceIndex = 0;
+    {
+      // Count rendered host nodes by traversing `returnFiber` subtree until `fiber` is found.
+      let node = returnFiber.child;
+      const nextNodeStack = [];
+      while (node && node !== fiber) {
+        if (node.tag === HostComponent || node.tag === HostText) {
+          ++hydrationWarningHostInstanceIndex;
```

Comment:
If a node is a host component, why do we need to go *inside* of it? I imagine we'd want to skip over it to search for the next one. Otherwise we'd count its children (which we shouldn't). Am I wrong?

Reply:
Yes I just had this thought, too, while drawing the diagram. Should it not have children by definition though?

Reply:
Why not?

Reply:
Okay, I think I got it: if a HostComponent is e.g. a div, it may have children. Then we should only count children-less HostComponents and HostTexts.

Reply:
Why shouldn’t we couldn’t them? 

I think we should count them regardless of whether they have children or not. We just shouldn’t descend *into* them because *their* children have no effect on the index. 

That said see my comment below.